### PR TITLE
[CF-165] Add scenario check action

### PR DESCRIPTION
**Pull request by mirrorcoder**
Added check for file existence.
Added check the correct of the struct.
Added check for stages.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/scheduler/scenario.py [hunk 1]
```diff
@@ -57,8 +61,11 @@ def init_tasks(self, init):
     def load_scenario(self, path_scenario=None):
         if path_scenario is None:
             path_scenario = self.path_scenario
+        if not os.path.isfile(path_scenario):
+            raise NotFoundScenarioFileError(path_scenario)
```
Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: This check is repeated twice, I'd suggest that you isolate load of scenario file and raise this error only if file cannot be opened for read.

File: cloudferrylib/scheduler/scenario.py [hunk 2]
```diff
@@ -176,3 +194,27 @@ def __init__(self, scenario_checker):
...
+class NotFoundScenarioFileError(RuntimeError):
+    def __init__(self, path):
+        super(NotFoundScenarioFileError, self).__init__(
+            "Not found on path %s" % path)
+
+
+class NotFoundTasksFileError(RuntimeError):
+    def __init__(self, path):
+        super(NotFoundTasksFileError, self).__init__(
+            "Not found on path %s" % path)
```
Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: None of these exceptions provide enough information to the user, neither does the calling code. What I'd suggest is for  incorrect scenario to provide the file and line number where the error was spotted, with some sort of clue to how to resolve the problem. For the "file not found" types of errors you should provide more information, like "Make sure you specified correct config value <config option> in <config file name>".

The other thing to improve is to have some better structure for exceptions and handling those appropriately, for example, "invalid scenario" exception is most likely a syntax error, while "file not found" exceptions are "invalid configuration" errors, those need to be caught and processed appropriately in the calling code.

File: cloudferrylib/scheduler/scenario.py (hunk 1 above)
Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: I didn't understoond, what do you mean?

File: cloudferrylib/scheduler/scenario.py [hunk 3]
```diff
@@ -57,8 +61,11 @@ def init_tasks(self, init):
     def load_scenario(self, path_scenario=None):
         if path_scenario is None:
             path_scenario = self.path_scenario
+        if not os.path.isfile(path_scenario):
+            raise NotFoundScenarioFileError(path_scenario)
         with open(path_scenario) as scenario_file:
             migrate = yaml.load(scenario_file)
+            ScenarioChecker.check_structure(migrate)
```
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: `self.check_structure(migrate)`

Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: You implemented that method in the same class and you do not need to call the method using a name of the class. Python allows you to call the method using `self`

File: cloudferrylib/scheduler/scenario.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and#
 # limitations under the License.
 
-
+import os.path
```
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: `import os` is enough

File: cloudferrylib/scheduler/scenario.py [hunk 5]
```diff
@@ -108,6 +115,17 @@ def __init__(self, tasks, key, scenario):
         self.links = set()
         self.active_tasks = set()
 
+    @staticmethod
+    def check_structure(body):
+        if not body or type(body) is not dict:
```
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: `type(body) is not dict` -> `not isinstance(body, dict)`
Probably in the future we will have a special class inherited from the dict to store the configuration and we will get the problem with this unit test

File: cloudferrylib/scheduler/scenario.py [hunk 6]
```diff
@@ -108,6 +115,17 @@ def __init__(self, tasks, key, scenario):
...
+            raise IncorrectScenarioStructure()
+        exist_one_of_stages = False
+        for stage in STAGES_SCENARIO:
+            if stage in body:
+                exist_one_of_stages = True
```
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: just `return`, and you will not have to have the `exist_one_of_stages` variable

File: cloudferrylib/scheduler/scenario.py [hunk 7]
```diff
@@ -108,6 +115,17 @@ def __init__(self, tasks, key, scenario):
         self.links = set()
         self.active_tasks = set()
 
+    @staticmethod
+    def check_structure(body):
+        if not body or type(body) is not dict:
+            raise IncorrectScenarioStructure()
+        exist_one_of_stages = False
+        for stage in STAGES_SCENARIO:
```
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: What do you think about this:
```python
STAGES_SCENARIO = {'migration', 'preparation', 'rollback'}
...
stages = set(body.keys)
if STAGES_SCENARIO.isdisjoint(stages):
    raise NoStagesScenario()
```

Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: Interesting

File: cloudferrylib/scheduler/scenario.py (hunk 5 above)
Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: i understand

File: cloudferrylib/scheduler/scenario.py (hunk 3 above)
Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: i didn't understoond

Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: i mean that

Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: i'm sorry, It's implemented in another class.

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.scheduler.scenario
cloudferrylib/scheduler/scenario.py:62: [W0201(attribute-defined-outside-init), 
Scenario.load_scenario_with_num_lines] Attribute 'raw_scenario_file' defined outside 
__init__
cloudferrylib/scheduler/scenario.py:64: [W0201(attribute-defined-outside-init), 
Scenario.load_scenario_with_num_lines] Attribute 'raw_scenario_file' defined outside 
__init__

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/scheduler/scenario.py [hunk 8]
```diff
@@ -68,13 +81,15 @@ def load_scenario(self, path_scenario=None):
...
 
     def get_net(self):
         result = {}
-        for key in ['migration', 'preparation', 'rollback']:
+        for key in STAGES_SCENARIO:
             if hasattr(self, key) and getattr(self, key):
                 scenario = getattr(self, key)
-                checker = ScenarioChecker(self.tasks.keys(), key, scenario)
+                checker = ScenarioChecker(self.tasks.keys(), key, scenario,
+                                          self.raw_scenario_file)
```
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: Why do we need to analyze the raw scenario file? Is the information stored in `self.preparation`, `self.rollback`, `self.namespace` and `self.migration` not enough? 
Please revert the patch to previous version and resolve all comments.

**Comment by SVilgelm** [comment]
:-1: The patch has lots of duplicated logic.

File: cloudferrylib/scheduler/scenario.py (hunk 8 above)
Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: no

Comment (mirrorcoder) [inline cloudferrylib/scheduler/scenario.py]: But may be you have other idea?

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by roman-verchikov** [comment]
-1, looks like this patch leads to failures in 33 functional tests, these need  to be resolved

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloud.os2os
cloud/os2os.py:16: [W0403(relative-import), ] Relative import 'cloud', should be 
'cloud.cloud'
cloud/os2os.py:17: [W0403(relative-import), ] Relative import 'cloud_ferry', should be 
'cloud.cloud_ferry'
************* Module cloudferrylib.scheduler.scheduler
cloudferrylib/scheduler/scheduler.py:25: [W0403(relative-import), ] Relative import 
'cursor', should be 'cloudferrylib.scheduler.cursor'
cloudferrylib/scheduler/scheduler.py:26: [W0403(relative-import), ] Relative import 
'signal_handler', should be 'cloudferrylib.scheduler.signal_handler'
cloudferrylib/scheduler/scheduler.py:27: [W0403(relative-import), ] Relative import 
'task', should be 'cloudferrylib.scheduler.task'
cloudferrylib/scheduler/scheduler.py:28: [W0403(relative-import), ] Relative import 
'thread_tasks', should be 'cloudferrylib.scheduler.thread_tasks'
cloudferrylib/scheduler/scheduler.py:60: [W0613(unused-argument), 
BaseScheduler.event_error_task] Unused argument 'task'
cloudferrylib/scheduler/scheduler.py:60: [W0613(unused-argument), 
BaseScheduler.event_error_task] Unused argument 'e'
cloudferrylib/scheduler/scheduler.py:78: [W0703(broad-except), 
BaseScheduler.process_chain] Catching too general exception Exception
cloudferrylib/scheduler/scheduler.py:87: [W0201(attribute-defined-outside-init), 
BaseScheduler.process_chain] Attribute 'exception' defined outside __init__
cloudferrylib/scheduler/scheduler.py:110: [W0201(attribute-defined-outside-init), 
BaseScheduler.addCursor] Attribute 'cursor' defined outside __init__

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3150/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloud.os2os
cloud/os2os.py:16: [W0403(relative-import), ] Relative import 'cloud', should be 
'cloud.cloud'
cloud/os2os.py:17: [W0403(relative-import), ] Relative import 'cloud_ferry', should be 
'cloud.cloud_ferry'
************* Module tests.scheduler.test_scheduler
tests/scheduler/test_scheduler.py:36: [E1136(unsubscriptable-object), 
SchedulerTestCase.test_start_scheduler] Value 's.event_start_task.call_args' is 
unsubscriptable
tests/scheduler/test_scheduler.py:38: [E1136(unsubscriptable-object), 
SchedulerTestCase.test_start_scheduler] Value 's.event_end_task.call_args' is 
unsubscriptable
************* Module cloudferrylib.scheduler.scheduler
cloudferrylib/scheduler/scheduler.py:25: [W0403(relative-import), ] Relative import 
'cursor', should be 'cloudferrylib.scheduler.cursor'
cloudferrylib/scheduler/scheduler.py:26: [W0403(relative-import), ] Relative import 
'signal_handler', should be 'cloudferrylib.scheduler.signal_handler'
cloudferrylib/scheduler/scheduler.py:27: [W0403(relative-import), ] Relative import 
'task', should be 'cloudferrylib.scheduler.task'
cloudferrylib/scheduler/scheduler.py:28: [W0403(relative-import), ] Relative import 
'thread_tasks', should be 'cloudferrylib.scheduler.thread_tasks'
cloudferrylib/scheduler/scheduler.py:60: [W0613(unused-argument), 
BaseScheduler.event_error_task] Unused argument 'task'
cloudferrylib/scheduler/scheduler.py:60: [W0613(unused-argument), 
BaseScheduler.event_error_task] Unused argument 'e'
cloudferrylib/scheduler/scheduler.py:78: [W0703(broad-except), 
BaseScheduler.process_chain] Catching too general exception Exception
cloudferrylib/scheduler/scheduler.py:87: [W0201(attribute-defined-outside-init), 
BaseScheduler.process_chain] Attribute 'exception' defined outside __init__
cloudferrylib/scheduler/scheduler.py:110: [W0201(attribute-defined-outside-init), 
BaseScheduler.addCursor] Attribute 'cursor' defined outside __init__

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloud.os2os
cloud/os2os.py:16: [W0403(relative-import), ] Relative import 'cloud', should be 
'cloud.cloud'
cloud/os2os.py:17: [W0403(relative-import), ] Relative import 'cloud_ferry', should be 
'cloud.cloud_ferry'
************* Module cloudferrylib.scheduler.scheduler
cloudferrylib/scheduler/scheduler.py:25: [W0403(relative-import), ] Relative import 
'cursor', should be 'cloudferrylib.scheduler.cursor'
cloudferrylib/scheduler/scheduler.py:26: [W0403(relative-import), ] Relative import 
'signal_handler', should be 'cloudferrylib.scheduler.signal_handler'
cloudferrylib/scheduler/scheduler.py:27: [W0403(relative-import), ] Relative import 
'task', should be 'cloudferrylib.scheduler.task'
cloudferrylib/scheduler/scheduler.py:28: [W0403(relative-import), ] Relative import 
'thread_tasks', should be 'cloudferrylib.scheduler.thread_tasks'
cloudferrylib/scheduler/scheduler.py:60: [W0613(unused-argument), 
BaseScheduler.event_error_task] Unused argument 'task'
cloudferrylib/scheduler/scheduler.py:60: [W0613(unused-argument), 
BaseScheduler.event_error_task] Unused argument 'e'
cloudferrylib/scheduler/scheduler.py:78: [W0703(broad-except), 
BaseScheduler.process_chain] Catching too general exception Exception
cloudferrylib/scheduler/scheduler.py:87: [W0201(attribute-defined-outside-init), 
BaseScheduler.process_chain] Attribute 'exception' defined outside __init__
cloudferrylib/scheduler/scheduler.py:110: [W0201(attribute-defined-outside-init), 
BaseScheduler.addCursor] Attribute 'cursor' defined outside __init__
************* Module tests.scheduler.test_scheduler
tests/scheduler/test_scheduler.py:36: [E1136(unsubscriptable-object), 
SchedulerTestCase.test_start_scheduler] Value 's.event_start_task.call_args' is 
unsubscriptable
tests/scheduler/test_scheduler.py:38: [E1136(unsubscriptable-object), 
SchedulerTestCase.test_start_scheduler] Value 's.event_end_task.call_args' is 
unsubscriptable

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
the renaming `migration` to `process` looks unnecessary

File: cloudferrylib/scheduler/scenario.py (hunk 8 above)
Comment (SVilgelm) [inline cloudferrylib/scheduler/scenario.py]: What do your 'no' mean, please explain, what information is missing?

**Comment by mirrorcoder** [comment]
@SVilgelm Different names of one section in the code and in the fileю Is this the norm?

**Comment by SVilgelm** [comment]
> migration to process
got it, thanks.
Please rebase

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.scheduler.test_scheduler
tests/scheduler/test_scheduler.py:36: [E1136(unsubscriptable-object), 
SchedulerTestCase.test_start_scheduler] Value 's.event_start_task.call_args' is 
unsubscriptable
tests/scheduler/test_scheduler.py:38: [E1136(unsubscriptable-object), 
SchedulerTestCase.test_start_scheduler] Value 's.event_end_task.call_args' is 
unsubscriptable

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/scheduler/scenario.py [hunk 9]
```diff
@@ -119,18 +157,21 @@ def check(self):
         throw_exception = False
         # log all errors if needed
         if self.missed_links:
-            LOG.error("Scenario has missed links: %s",
-                      ", ".join(self.missed_links))
+            lines = self.get_lines_tasks(self.missed_links)
+            LOG.error("Scenario has missed links: \n %s",
```
Comment (roman-verchikov) [inline cloudferry/lib/scheduler/scenario.py]: "missing" not "missed"

**Comment by roman-verchikov** [comment]
Closed due to low priority

**Pull request closed without merging**